src/Uno.UWP/Devices/Geolocation/Geolocator.Android.cs

```diff
@@ -11,113 +11,334 @@
-			}
+        MyLocListen myListener = new MyLocListen();
+        #endregion 
 
-			return locationManager;
-		}
+        public Geolocator()
+        {
+            _locationManager = (LocationManager)Android.App.Application.Context.GetSystemService(Android.Content.Context.LocationService);
 
-		partial void StartPositionChanged()
-		{
+            _reportInterval = 1000;
+            _movementThreshold = 0;
+
+            RequestUpdates();
+        }
+
+        ~Geolocator()
+        {
+            try
+            {
+                RemoveUpdates();
+            }
+            catch
+            { }
+
+            try
+            {
+                _locationManager.Dispose();
+            }
+            catch
+            { }
+        }
+
+        public Task<Windows.Devices.Geolocation.Geoposition> GetGeopositionAsync()
+            => GetGeopositionAsync(TimeSpan.FromHours(2), TimeSpan.FromSeconds(60));
+
+        public async Task<Windows.Devices.Geolocation.Geoposition> GetGeopositionAsync(TimeSpan maximumAge, TimeSpan timeout)
+        {
+            var providers = _locationManager.GetProviders(_locationCriteria,true);
+            int bestAccuracy = 10000;
+            Location bestLocation = null;
+			
 			BroadcastStatus(PositionStatus.Initializing);
-		}
 
-		public void OnLocationChanged(Location location)
-		{
-			BroadcastStatus(PositionStatus.Ready);
-			this._positionChanged?.Invoke(this, new PositionChangedEventArgs(location.ToGeoPosition()));
-		}
+            foreach (string locationProvider in providers)
+            {
+                var location = _locationManager.GetLastKnownLocation(locationProvider);
+                if (location != null)
+                {
+                    // check how old is this fix
+                    DateTimeOffset date = DateTimeOffset.FromUnixTimeMilliseconds(location.Time); 
+                    if (date.AddMilliseconds(maximumAge.TotalMilliseconds) > DateTimeOffset.UtcNow)
+                    {   // can be used, but is it best accuracy?
+                        if (location.HasAccuracy)
+                        {
+                            if (location.Accuracy < bestAccuracy)
+                            {
+                                bestAccuracy = (int)location.Accuracy;
+                                bestLocation = location;
+                            }
+                        }
+                        else
+                            bestLocation = location;
+                    }
+                }
+            }
 
-		public void OnProviderDisabled(string provider)
-		{
-		}
+            if (bestLocation != null)
+            {
+                RemoveUpdates();
+				BroadcastStatus(PositionStatus.Ready);
+                return bestLocation.ToGeoPosition();
+            }
 
-		public void OnProviderEnabled(string provider)
-		{
-		}
+            // wait for fix
+            for (int i = (int)(timeout.TotalMilliseconds / 250.0); i > 0; i--)
```

Comment:
Not a big problem, but if the CPU is very busy, you have no warranty your `Task.Delay()` will take exactly 250ms.  So your timeout could drift here.

Instead, you should simply calculate expiration before the loop and test if time is out each time.

Reply:
I think this is almost the same (my version and version proposed by you) - 'tick' is 250 ms, so if someone defines timeout e.g. 200 ms, it would be 250 ms, and 300 ms became 500 ms. So, even 50 ms of drift would be imperceptible.